File: py/src/rulecat/rule.py
from __future__ import annotations
from typing import TYPE_CHECKING, Any

from rulecat.rule_id import rule_id

if TYPE_CHECKING:
    from rulecat.graph import Graph
    from rulecat.types import Spec


class Rule:
    def __init__(self, id: str, graph: Graph, idx: int, fn: Any) -> None:
        self.id = id
        self._graph = graph
        self._idx = idx
        self._fn = fn

    def attacks(self, target: Rule) -> None:
        from rulecat.defeat_edge import DefeatEdge
        self._graph.defeats.append(DefeatEdge(from_=self.id, to=target.id))

    def with_spec(self, spec: Spec) -> Rule:
        spec.validate()
        meta = self._graph.rules[self._idx]
        meta.specs.append(spec)
        old_id = meta.name
        new_id = rule_id(self._fn, meta.specs)
        meta.name = new_id
        self.id = new_id
        role = self._graph.roles.pop(old_id, None)
        if role is not None:
            self._graph.roles[new_id] = role
        for edge in self._graph.defeats:
            if edge.from_ == old_id:
                edge.from_ = new_id
            if edge.to == old_id:
                edge.to = new_id
        return self

    def qualifier(self, q: float) -> Rule:
        if q < 0.0 or q > 1.0:
            raise ValueError("qualifier must be between 0.0 and 1.0")
        self._graph.rules[self._idx].qualifier = q
        return self

File: ts/src/rule.ts
import type { Spec, NodeHandler } from "./types.js";
import type { Graph } from "./graph.js";
import type { RuleFunc } from "./types.js";
import { ruleID } from "./rule-id.js";

export class Rule {
  id: string;
  private graph: Graph;
  private idx: number;
  private fn: RuleFunc;

  constructor(id: string, graph: Graph, idx: number, fn: RuleFunc) {
    this.id = id; this.graph = graph; this.idx = idx; this.fn = fn;
  }

  attacks(target: Rule): void {
    this.graph.defeats.push({ from: this.id, to: target.id });
  }

  with(spec: Spec): Rule {
    spec.validate();
    const meta = this.graph.rules[this.idx];
    meta.specs.push(spec);
    const oldID = meta.name;
    const newID = ruleID(this.fn, meta.specs);
    meta.name = newID;
    this.id = newID;
    const role = this.graph.roles.get(oldID);
    if (role !== undefined) {
      this.graph.roles.delete(oldID);
      this.graph.roles.set(newID, role);
    }
    for (const edge of this.graph.defeats) {
      if (edge.from === oldID) edge.from = newID;
      if (edge.to === oldID) edge.to = newID;
    }
    return this;
  }

  qualifier(q: number): Rule {
    if (q < 0.0 || q > 1.0) throw new Error("qualifier must be between 0.0 and 1.0");
    this.graph.rules[this.idx].qualifier = q;
    return this;
  }

  onActive(h: NodeHandler): Rule {
    this.graph.rules[this.idx].onActive = h;
    return this;
  }

  onDefeated(h: NodeHandler): Rule {
    this.graph.rules[this.idx].onDefeated = h;
    return this;
  }

  onInactive(h: NodeHandler): Rule {
    this.graph.rules[this.idx].onInactive = h;
    return this;
  }
}
